### PR TITLE
Variable elasticity tensor and stress class modifications

### PR DESCRIPTION
**Pull request by sapitts**
### Relevant Design Changes

The objective of this PR is to allow the standard isotropic stress calculators in tensor mechanics to handle the case when the elastic constants change during a simulations (e.g. as a function of temperature). To this end, we've added a new boolean material property, `elasticity_tensor_has_changed` to elasticity tensor classes which have variable isotropic elasticity constants.  Developers will be responsible for adding this `elasticity_tensor_has_changed` material property to new variable elasticity tensor classes.

The isotropic stress classes look for the `elasticity_tensor_has_changed` property with `getDefaultMaterialProperty`, and, if the boolean is true, these stress classes have been changed to calculate the old stress state as a function of the new elasticity tensor and the old strain.  The old variable elasticity tensor stress class has been deprecated.

This change does add a stateful rank-2 tensor `elastic_strain_old`; I'd appreciate advice on how to only add this material property if the elasticity tensor is changing.

### Test Cases
The existing varying isotropic elasticity tensor test files have been updated to use the regular `ComputeFiniteStrainElasticStress` class, and the deprecated class will remain with the `allow_deprecated_until` option for a month.

Refs #8616

### CODE REVIEW

**Comment by sapitts** [comment]
@bwspenc, @dschwen, and @WilkAndy would some combination of you guys review this merge request as you have time please?

**Comment by dschwen** [comment]
Wait, is that what we agreed on? I thought the more sensible approach would be to formulate the property as `elasticity_tensor_is_unchanged`. The `getDefaultMaterialProperty` call should return `false` by default. So if the user is not actively enabling the calculation shortcut he/she will at least not get a wrong result, only a slower claculation!

**Review by dschwen — CHANGES_REQUESTED** [review]
Needs discussion

**Comment by sapitts** [comment]
So.  I agree with @dschwen  about setting up the code to use the old strain, rather than the old stress, by default; however, I'm running into another problem with some of @WilkAndy's tests that have an initial stress provided to [ComputeStressBase](https://github.com/idaholab/moose/blob/devel/modules/tensor_mechanics/src/materials/ComputeStressBase.C#L18).

Is there a more elegant way to avoid to guard against this case than the super long if statement:
`if (!_elasticity_tensor_is_constant[_qp] && isElasticityTensorGuaranteedIsotropic() && !(isParamValid("initial_stress") )` ?

**Comment by sapitts** [comment]
@dschwen Changing the material property flag (which I've named `elasticity_tensor_is_constant` to avoid the confusion of the double negative) has turned this commit into a complete mess, and I'm no longer convinced that using the old strain to calculate the new/current stress by default is the 'right' thing to do.

I've put in a check in `initQpStatefulProperties` for the cases when both an initial stress is provided and a (potentially) variable elasticity tensor is used to error out.  This check actually makes physical sense: if the elasticity tensor is changing, then an initial strain should be applied instead of an initial stress since the stress will change with the elasticity tensor.

To give users an elasticity tensor option for test cases using an initial stress, I've set the material property `elasticity_tensor_is_constant` in `ComputeIsotropicElasticityTensor` to true because this is the only class in which it makes a remote amount of sense to assume constant elasticity tensor constants.

There are two stumbling blocks with this approach that assumes that all elasticity tensors have changed by default:
 1. In setting the `ComputeIsotropicElasticityTensor` material property `elasticity_tensor_is_constant = true`, I'm setting new users up for failure in that a new variable elasticity tensor is likely to inherit from `ComputeIsotropicElasticityTensor`. In this case, we've got the same problem that the user has to know to set the material property `elasticity_tensor_is_constant` as from my first attempt.
2. The Cosserat classes that @WilkAndy  has developed assume that the new stress is calculated with the old stress.  This current commit will fail in the coal mining example problem because of the change to use the old strain to calculate the new stress.

In summary, assuming that all elasticity tensors are variable by default does not eliminate the chance that a user will create a variable elasticity tensor with the wrong `elasticity_tensor_is_constant` value, but this approach sure does mess up @WilkAndy's Cosserat work.

**Comment by sapitts** [comment]
@WilkAndy I was unsuccessful in convincing @dschwen to let me assume that the elasticity tensors are constant by default, so, could I please ask you for some help with sorting out the Cosserat and Capped weak plane/ Drucker-Prager classes that run calculations with an old stress value?  I read Forrest et al (1997) once, but it's been a while, and I'm not even sure if that's the same approach that you've used here.  I've separated out the changes I made to the cosserat specific input files and classes to maximize the number of passing classes into the third commit.

@idaholab/moose-team This approach uses a material property at each qp point when all we really need is a single yes/no once per material, like the `isElasticityTensorGuaranteedIsotropic()`. Is there an interface/ proper direction for a limited coder like myself to follow to create a `isElasticityTensorGuaranteedConstant()` type function?

File: modules/tensor_mechanics/src/materials/ComputeMultipleInelasticCosseratStress.C [hunk 1]
```diff
@@ -35,7 +35,8 @@ ComputeMultipleInelasticCosseratStress::ComputeMultipleInelasticCosseratStress(
 void
 ComputeMultipleInelasticCosseratStress::initQpStatefulProperties()
 {
-  ComputeMultipleInelasticStress::initQpStatefulProperties();
```
Comment (WilkAndy) [inline modules/tensor_mechanics/src/materials/ComputeMultipleInelasticCosseratStress.C]: Why did you make this change?

File: modules/tensor_mechanics/include/materials/ComputeIsotropicElasticityTensor.h [hunk 2]
```diff
@@ -38,6 +37,9 @@ class ComputeIsotropicElasticityTensor : public ComputeElasticityTensorBase
 
   /// Individual elasticity tensor
   RankFourTensor _Cijkl;
+
+  /// Flag for variable elasticity isotropic tensors
+  MaterialProperty<bool> & _elasticity_tensor_is_constant;
```
Comment (WilkAndy) [inline modules/tensor_mechanics/include/materials/ComputeIsotropicElasticityTensor.h]: i agree this would be better not to be a MaterialProperty - just a single bool would be better.

File: modules/tensor_mechanics/src/materials/ComputeFiniteStrainElasticStress.C [hunk 3]
```diff
@@ -34,11 +37,28 @@ ComputeFiniteStrainElasticStress::initialSetup()
...
 
 void
+ComputeFiniteStrainElasticStress::initQpStatefulProperties()
+{
+  ComputeStressBase::initQpStatefulProperties();
+
+  if ((isParamValid("initial_stress")) && !_elasticity_tensor_is_constant[_qp])
+    mooseError("A finite stress material cannot both have an initial stress and an elasticity "
+               "tensor with varying values; please use a defined constant elasticity tensor, "
+               "such as ComputeIsotropicElasticityTensor, if your model defines an initial stress");
```
Comment (WilkAndy) [inline modules/tensor_mechanics/src/materials/ComputeFiniteStrainElasticStress.C]: You could add "or apply an initial strain" to this message

File: modules/tensor_mechanics/src/materials/ComputeFiniteStrainElasticStress.C [hunk 4]
```diff
@@ -34,11 +37,28 @@ ComputeFiniteStrainElasticStress::initialSetup()
...
+void
 ComputeFiniteStrainElasticStress::computeQpStress()
 {
   // Calculate the stress in the intermediate configuration
-  RankTwoTensor intermediate_stress =
-      _stress_old[_qp] + _elasticity_tensor[_qp] * _strain_increment[_qp];
+  RankTwoTensor intermediate_stress;
+
+  // Check if the elasticity tensor has changed values
+  if (!_elasticity_tensor_is_constant[_qp])
```
Comment (WilkAndy) [inline modules/tensor_mechanics/src/materials/ComputeFiniteStrainElasticStress.C]: Here and elsewhere: do we want this check?  why not just always use stress = E * (strain_old + strain_increment) ?

**Review by WilkAndy — CHANGES_REQUESTED** [review]

File: modules/tensor_mechanics/src/materials/ComputeFiniteStrainElasticStress.C (hunk 4 above)
Comment (WilkAndy) [inline modules/tensor_mechanics/src/materials/ComputeFiniteStrainElasticStress.C]: Oh, i think this is because of initial_stress - right?  Otherwise you'd get stress=0 at the start of the simulation, even if you used initial_stress = something.

**Comment by WilkAndy** [comment]
ComputeMultipleInelasticCosseratStress.  I believe that no changes need to be made, assuming you put back the call to ComputeMultipleInelasticStress::initQpProperties, that has that check of constant E in it (then the coal_mining example should not produce an exodiff, as at present i think the diff is due to the code not actually applying the initial stresses).  There are Cosserat equivalents of stress = E * strain  (```_couple_stress[_qp] = _elastic_flexural_rigidity_tensor[_qp] * _curvature[_qp];```) but at the moment i don't think there's a way for a user to make the flexural rigidity tensor non-constant.

ComputeLayeredCosseratElasticityTensor  should have ```isElasticityTensorGuaranteedConstant``` set true

ComputeCosseratElasticityTensor  should probably have the flag set true

**Comment by WilkAndy** [comment]
Overall i'd like this to be generalised to anisotropic E, as constantness and isotropy are completely unrelated.  You know that i simulated primarily using Cosserat mechanics which is, for me, always anisotropic.   At the moment i do have constant E, but who knows what the future holds.  Why not put this flag in ComputeElasticityTensorBase ?  This current PR fractures TensorMechanics into several distinct parts which is unappetizing to me.

I also think the flag should default to true, as it is the most common use case.  A future developer of a non-constant elasticity class might get stung by this (because they forget to override the "true") but it'd be only a brief moment of pain while they ran their first test case and wondered why their results were crappy.  In reality though, when people override classes they usually check the header file to see what they should override.

File: modules/tensor_mechanics/src/materials/ComputeFiniteStrainElasticStress.C (hunk 4 above)
Comment (sapitts) [inline modules/tensor_mechanics/src/materials/ComputeFiniteStrainElasticStress.C]: yes, fortunately your tests helped me catch the initial stress cases

File: modules/tensor_mechanics/include/materials/ComputeIsotropicElasticityTensor.h (hunk 2 above)
Comment (sapitts) [inline modules/tensor_mechanics/include/materials/ComputeIsotropicElasticityTensor.h]: I talked with @permcody about the single bool option, and he recommended just using a material property because there isn't a good interface in place.

File: modules/tensor_mechanics/src/materials/ComputeMultipleInelasticCosseratStress.C (hunk 1 above)
Comment (sapitts) [inline modules/tensor_mechanics/src/materials/ComputeMultipleInelasticCosseratStress.C]: I added this flag to skip over the check for a constant elasticity tensor flag that I added into ComputeMultipleInelasticStress.  With the inheritance as it initially was, I was seeing test failures in several of the jacobian tests (cto29, cwpc01, cwpc02, cdpc01, cdpc02, cdp_cwp_coss01 & 02, coss_elastic) and csv diffs in the capped_weak_plane/small_cosserat tests.  I hadn't tried adding the flag to ComputeCosseratElasticityTensor though, so I'll give that a try.

**Comment by sapitts** [comment]
@WilkAndy Thank you for all of your suggestions on how to get this commit to work.  In this case we have to keep both the check for constant and isotropic, , if the material is not isotropic, we cannot correctly calculate the old stress value from the old strain. 

@bwspenc, Andy makes another good point for the case of having the default assumption be that the elasticity tensor does NOT change in that the standard use case is for a constant elasticity tensor and the use of the old_stress.  Where do you side on this issue?

File: modules/tensor_mechanics/include/materials/ComputeIsotropicElasticityTensor.h (hunk 2 above)
Comment (WilkAndy) [inline modules/tensor_mechanics/include/materials/ComputeIsotropicElasticityTensor.h]: But isn't the isIsotropic thing basically a single bool?

File: modules/tensor_mechanics/src/materials/ComputeMultipleInelasticCosseratStress.C (hunk 1 above)
Comment (WilkAndy) [inline modules/tensor_mechanics/src/materials/ComputeMultipleInelasticCosseratStress.C]: Please could you make ```ComputeMultipleInelasticCosseratStress::initQpStatefulProperties()``` call ```ComputeMultipleInelasticStress::initQpStatefulProperties()```, as then when changes get made in the latter (such as this PR!) they are automatically picked up in the Cosserat case.

**Comment by WilkAndy** [comment]
@sapitts , could you explain "if the material is not isotropic, we cannot correctly calculate the old stress value from the old strain." ?

**Comment by sapitts** [comment]
Yes, or at least I'll get stuck trying; that code is just on a different
computer at work so I haven't made the change yet

On Sun, Jul 9, 2017 at 6:29 PM, Andy Wilkins <notifications@github.com>
wrote:

> *@WilkAndy* commented on this pull request.
> ------------------------------
>
> In modules/tensor_mechanics/src/materials/ComputeMultipleInelasticCosser
> atStress.C
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_idaholab_moose_pull_9438-23discussion-5Fr126320038&d=DwMCaQ&c=54IZrppPQZKX9mLzcGdPfFD1hxrcB__aEkJFOKJFd00&r=FAWlXtgI_7SSz_WkuPhGTv9GpGZquhqj8j2e4CHgrB0&m=kGsOkIx0Qex12icSoQMJ3ynKaPvbLaW4wNJKM2_tmP8&s=D65vHJJbqanJrzspieivc_xkjTQWVLTt5PSTTbuXANQ&e=>
> :
>
> > @@ -35,7 +35,8 @@ ComputeMultipleInelasticCosseratStress::ComputeMultipleInelasticCosseratStress(
>  void
>  ComputeMultipleInelasticCosseratStress::initQpStatefulProperties()
>  {
> -  ComputeMultipleInelasticStress::initQpStatefulProperties();
>
> Please could you make ComputeMultipleInelasticCosseratStress::
> initQpStatefulProperties() call ComputeMultipleInelasticStress
> ::initQpStatefulProperties(), as then when changes get made in the latter
> (such as this PR!) they are automatically picked up in the Cosserat case.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_idaholab_moose_pull_9438-23discussion-5Fr126320038&d=DwMCaQ&c=54IZrppPQZKX9mLzcGdPfFD1hxrcB__aEkJFOKJFd00&r=FAWlXtgI_7SSz_WkuPhGTv9GpGZquhqj8j2e4CHgrB0&m=kGsOkIx0Qex12icSoQMJ3ynKaPvbLaW4wNJKM2_tmP8&s=D65vHJJbqanJrzspieivc_xkjTQWVLTt5PSTTbuXANQ&e=>,
> or mute the thread
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AI6P5UnTfHwdj3UzZW3W-2DK8UFIYokwP8ks5sMXBqgaJpZM4OMkrA&d=DwMCaQ&c=54IZrppPQZKX9mLzcGdPfFD1hxrcB__aEkJFOKJFd00&r=FAWlXtgI_7SSz_WkuPhGTv9GpGZquhqj8j2e4CHgrB0&m=kGsOkIx0Qex12icSoQMJ3ynKaPvbLaW4wNJKM2_tmP8&s=0zcUWzUPh-2SuK-sGm8p4WacnA4D0AxKsa5hOoOHXFY&e=>
> .
>



-- 
Stephanie Pitts
PhD Candidate, Mechanical Engineering
Washington State University
Intern, Fuel Modeling and Simulation
Idaho National Laboratory

**Comment by sapitts** [comment]
Ben might have to correct me, but I believe that we found that for
anisotropic elasticity tensors in bending, we had a bending test that
exodiffed from the original old stress method compared to the old strain
method.

On Sun, Jul 9, 2017 at 6:32 PM, Andy Wilkins <notifications@github.com>
wrote:

> @sapitts
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_sapitts&d=DwMCaQ&c=54IZrppPQZKX9mLzcGdPfFD1hxrcB__aEkJFOKJFd00&r=FAWlXtgI_7SSz_WkuPhGTv9GpGZquhqj8j2e4CHgrB0&m=K7YGgcd9WnCfV1nbbK4kzrc527wEXynqlWg7al3PPEk&s=G5zm1QB4vLaE-89Mbdu25z2EURE9GIs7d_-loU9W_Y8&e=>
> , could you explain "if the material is not isotropic, we cannot correctly
> calculate the old stress value from the old strain." ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_idaholab_moose_pull_9438-23issuecomment-2D313975986&d=DwMCaQ&c=54IZrppPQZKX9mLzcGdPfFD1hxrcB__aEkJFOKJFd00&r=FAWlXtgI_7SSz_WkuPhGTv9GpGZquhqj8j2e4CHgrB0&m=K7YGgcd9WnCfV1nbbK4kzrc527wEXynqlWg7al3PPEk&s=goA0t6vYmV4EMgLSiMSWbvalJ-hm5PwyfijqO31DEDw&e=>,
> or mute the thread
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AI6P5dv-5FfH6SAxuj544m9rRO2ay-5FaPWHks5sMXEfgaJpZM4OMkrA&d=DwMCaQ&c=54IZrppPQZKX9mLzcGdPfFD1hxrcB__aEkJFOKJFd00&r=FAWlXtgI_7SSz_WkuPhGTv9GpGZquhqj8j2e4CHgrB0&m=K7YGgcd9WnCfV1nbbK4kzrc527wEXynqlWg7al3PPEk&s=EvxpYGTmsGDZRMRUEMfwpIV924GDFEZDSieVuo1Nk8E&e=>
> .
>



-- 
Stephanie Pitts
PhD Candidate, Mechanical Engineering
Washington State University
Intern, Fuel Modeling and Simulation
Idaho National Laboratory

**Comment by WilkAndy** [comment]
Anisotropic old stress issue: thanks @sapitts , i think we need to get to the bottom of that.  I don't see any fundamental reason why stress_old = E * elastic_strain_old shouldn't always work.  Feel free to educate me!

**Comment by sapitts** [comment]
@WilkAndy There's a long discussion about the elasticity tensor and rotations (or lack thereof) in the isIsotropicGuarantee pull request #8881

File: modules/tensor_mechanics/src/materials/ComputeIsotropicElasticityTensor.C [hunk 5]
```diff
@@ -124,6 +126,9 @@ ComputeIsotropicElasticityTensor::ComputeIsotropicElasticityTensor(
 void
 ComputeIsotropicElasticityTensor::computeQpElasticityTensor()
 {
+  // Let the Stress Calculator classes know that this class only has constant values
+  _elasticity_tensor_is_constant[_qp] = true;
+
```
Comment (sapitts) [inline modules/tensor_mechanics/src/materials/ComputeIsotropicElasticityTensor.C]: After talking with @dschwen, I've moved this material property from initQpStatefulProperties, since it is not intended to be stateful anyway, to the main compute method to ensure that inheriting classes overwrite the setting of the `_elasticity_tensor_is_constant` flag.  It's definitely not the most computationally efficient place to set this value, so please let me know if there's somewhere else I should set the value.

File: modules/tensor_mechanics/src/materials/ComputeMultipleInelasticCosseratStress.C [hunk 6]
```diff
@@ -42,6 +42,9 @@ ComputeMultipleInelasticCosseratStress::initQpStatefulProperties()
 void
 ComputeMultipleInelasticCosseratStress::computeQpStress()
 {
+  if (!_elasticity_tensor_is_constant[_qp])
+    mooseError("The Cosserat stress classes assume constant elasticity tensors");
+
```
Comment (sapitts) [inline modules/tensor_mechanics/src/materials/ComputeMultipleInelasticCosseratStress.C]: @WilkAndy I've changed back the inheritance structure to `ComputeMultipleInelasticStress`, but I added in this check to keep users from trying to apply a variable elasticity tensor to the Cosserat code...

**Comment by WilkAndy** [comment]
Re: Isotropic and this PR.  In  #8881 I found

>The way it is currently implemented, ComputeFiniteStrainElasticStress only works with isotropic elasticity tensors because it does not rotate the elasticity tensor forward to the new configuration.

I see ```ComputeFiniteStrainElasticStress``` has the necessary ```isElasticityTensorGuaranteedIsotropic()```.  Good!  But that's a different issue.

I strongly believe this E-constantness should **not** be tied to isotropy, and the isotropic flag should be available to all classes deriving from ```ComputeElasticityTensorBase```.

This is because someone (like me) will be using elasticity tensors derived from ```ComputeElasticityTensorBase``` and not from ```ComputeIsotropicElasticityTensor```, so all the checks won't work properly.   But it's also unappealing to logically connect isotropy and E-constantness in the code just because of the separate issue of ```ComputeFiniteStrainElasticStress``` not correctly rotating E.

File: modules/tensor_mechanics/src/materials/ComputeMultipleInelasticCosseratStress.C (hunk 6 above)
Comment (WilkAndy) [inline modules/tensor_mechanics/src/materials/ComputeMultipleInelasticCosseratStress.C]: Looks fine.  Perhaps when things "shake down" a little in this PR this Cosserat-constant check might be (re)moved, i'm not sure.

**Comment by dschwen** [comment]
Constness and isotropy will not be tied together. I don't believe anyone suggested that. That would be a separate and independent guarantee (which is what I'd prefer over a material property, although it would be more limited, i.e. the same over space and time).

**Comment by WilkAndy** [comment]
Good, thanks @dschwen .   Then let's put the ```_elasticity_tensor_is_constant``` in ```ComputeElasticityTensorBase``` instead of in the derived classes.  I prefer the non-MaterialProperty too.

**Comment by dschwen** [comment]
@sapitts are you waiting for #9484 to use the new guarantees system to implement this?

**Comment by sapitts** [comment]
@dschwen, yes.  I think the is constant guarantee is the last thing I need to add to this commit to address the comments from you and @WilkAndy

File: modules/tensor_mechanics/src/materials/ComputeFiniteStrainElasticStress.C [hunk 7]
```diff
@@ -46,7 +46,7 @@ ComputeFiniteStrainElasticStress::initQpStatefulProperties()
 void
 ComputeFiniteStrainElasticStress::computeQpStress()
 {
-  if ((isParamValid("initial_stress")) && !_elasticity_tensor_is_constant[_qp])
+  if ((isParamValid("initial_stress")) && !_is_elasticity_tensor_guaranteed_constant_in_time)
```
Comment (dschwen) [inline modules/tensor_mechanics/src/materials/ComputeFiniteStrainElasticStress.C]: this check can and should now be moved to initial setup!

File: modules/tensor_mechanics/src/materials/ComputeMultipleInelasticStress.C [hunk 8]
```diff
@@ -108,23 +108,20 @@ ComputeMultipleInelasticStress::initialSetup()
...
-<<<<<<< HEAD
       if (rrr->requiresIsotropicTensor() && !_is_elasticity_tensor_guaranteed_isotropic)
-        mooseError("Model " + models[i] + " requires an isotropic elasticity tensor, but the one "
-                                          "supplied is not guaranteed isotropic");
-=======
-      if (rrr->requiresIsotropicTensor() && !isElasticityTensorGuaranteedIsotropic())
         mooseError("Model " + models[i] +
                    " requires an isotropic elasticity tensor, but the one supplied is not "
                    "guaranteed isotropic");
->>>>>>> Change old stress calculation based on elasticity_tensor_has_changed flag. Refs #8616
```
Comment (dschwen) [inline modules/tensor_mechanics/src/materials/ComputeMultipleInelasticStress.C]: whoops, what happened here?

File: modules/tensor_mechanics/src/materials/ComputeMultipleInelasticStress.C [hunk 9]
```diff
@@ -134,7 +131,7 @@ ComputeMultipleInelasticStress::initialSetup()
 void
 ComputeMultipleInelasticStress::computeQpStress()
 {
-  if ((isParamValid("initial_stress")) && !_elasticity_tensor_is_constant[_qp])
+  if ((isParamValid("initial_stress")) && !_is_elasticity_tensor_guaranteed_constant_in_time)
```
Comment (dschwen) [inline modules/tensor_mechanics/src/materials/ComputeMultipleInelasticStress.C]: This, too, should go to initial setup (the isParamValid check contains lots of slow string comparisons)

**Review by dschwen — CHANGES_REQUESTED** [review]

File: modules/tensor_mechanics/src/materials/ComputeFiniteStrainElasticStress.C (hunk 7 above)
Comment (sapitts) [inline modules/tensor_mechanics/src/materials/ComputeFiniteStrainElasticStress.C]: ah okay, I'll make that change.  thanks!

**Comment by sapitts** [comment]
@dschwen, @WilkAndy, and @bwspenc I have a merge request into Mastodon to add in the necessary is constant in time guarantee to one of the elasticity tensors; once that merge request is in, this one should be ready to go.  Would you take another look over this commit as is now and let me know if there additional changes you would like to see made or if I have addressed all of your comments?

**Review by dschwen — COMMENTED** [review]
This adds the conflict markers in the first commit and removes them in the second. I'd prefer if they never showed up in the first place.

**Comment by sapitts** [comment]
I'm squashing the commits to remove the git conflict markers from the history; thanks for catching that mistake (among the many others)

**Comment by dschwen** [comment]
I can squash on merge

**Review by dschwen — APPROVED** [review]

**Comment by sapitts** [comment]
well in that case thanks!

**Comment by WilkAndy** [comment]
I'll have a look at this next week !

a

**Comment by brianmoose** [comment]
As I am sure you aware since there is a big "PR: Failed but allowed" label, this broke a test in falcon.

**Comment by dschwen** [comment]
Oh. Let me see if I have access to that. Then I'll fix it this morning.

**Comment by dschwen** [comment]
I do have access. Working on it now.

**Comment by dschwen** [comment]
The breakage has nothing to do with this PR @brianmoose. Some Richards stuff got deprecated. Pinging @WilkAndy , maybe he knows how to fix this.

**Comment by brianmoose** [comment]
If you say so. I just noticed the pattern that falcon broke on every iteration of this PR, nowhere else. Broke on devel->master when this was merged and is now broken for other PRs.

**Comment by brianmoose** [comment]
Also, the error message is exactly an error introduced in this PR.....

**Comment by dschwen** [comment]
Yeah, I stopped reading after 50 deprecation messages :-D

#9517 should fix this!

File: modules/tensor_mechanics/include/materials/ComputeVariableIsotropicElasticityTensor.h [hunk 10]
```diff
@@ -30,6 +30,9 @@ class ComputeVariableIsotropicElasticityTensor : public ComputeElasticityTensorB
   /// Material defininig the Poisson's Ratio
   const MaterialProperty<Real> & _poissons_ratio;
 
+  /// Flag for variable elasticity isotropic tensors
+  MaterialProperty<bool> & _elasticity_tensor_is_constant;
```
Comment (WilkAndy) [inline modules/tensor_mechanics/include/materials/ComputeVariableIsotropicElasticityTensor.h]: is this ```_elasticity_tensor_is_constant``` in ```ComputeVariableIsotropicElasticityTensor``` still needed?

Comment (sapitts) [inline modules/tensor_mechanics/include/materials/ComputeVariableIsotropicElasticityTensor.h]: No :-( Thanks for the catch @WilkAndy

Comment (sapitts) [inline modules/tensor_mechanics/include/materials/ComputeVariableIsotropicElasticityTensor.h]: Corrected in #9521
